compiler/resolution/resolveFunction.cpp

```diff
@@ -123,6 +123,22 @@ static void resolveFormals(FnSymbol* fn) {
       }
     }
 
+    if (formal->type == dtVoid && !formal->hasFlag(FLAG_TYPE_VARIABLE)) {
+      Symbol* ps = formal->defPoint->parentSymbol;
+      if (ps) {
+        if (!ps->hasFlag(FLAG_COMPILER_GENERATED))
+          USR_FATAL(formal,
+                    "Formal '%s' cannot be declared 'void'."
+                    " Consider using 'nothing' instead.",
+                    formal->name);
+        else if (ps->name == astrInit)
+          USR_FATAL(formal,
+                    "Field '%s' cannot be declared 'void'."
+                    " Consider using 'nothing' instead.",
+                    formal->name);
+      }
+    }
+
```

Comment:
If the formal has `void` type, and its parent symbol is *not* compiler-generated, then it is an invalid `void` source-code formal.

If the formal has `void` type, and its parent symbol *is* compiler-generated, and the parent symbol is `astrInit`, then it is an aggregate field intialization, such as from `new c();` and the field cannot be `void`.